tests/test_testing/test_matcher_engine.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+
+import random
+import re
+
+import pytest
+
+from sphinx.testing.matcher import _engine as engine
+
+
+def test_line_pattern():
+    assert engine.to_line_patterns('a') == ('a',)
+    assert engine.to_line_patterns(['a', 'b']) == ('a', 'b')
+
+    p = re.compile('b')
+    assert engine.to_line_patterns(p) == (p,)
+    assert engine.to_line_patterns(['a', p]) == ('a', p)
+
+    # ensure build reproducibility
+    assert engine.to_line_patterns({'a', p}) == ('a', p)
+
+    p1 = re.compile('a')
+    p2 = re.compile('a', re.MULTILINE)
+    p3 = re.compile('ab')
+    p4 = re.compile('ab', re.MULTILINE)
+    ps = (p1, p2, p3, p4)
+
+    for _ in range(100):
```

Comment:
Do we anticipate that additional cases are likely to be added here?  (`4! = 24`, so if not, perhaps we should use `itertools` to generate the permutations instead)

Reply:
Err.... Actually no. Good catch ! (I think it's something I forgot to change).

Reply:
Ah ! I remember why I did this ! it's to be sure that sets are considered correctly (I mean, the type `set`). The reason is that I wanted to have multiple objects of the type `set` that are independent to each other.